File: typescript/src/tracing.ts
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

/**
 * Tracing helpers — mirrors Python provide.telemetry @trace decorator and tracer access.
 *
 * Uses @opentelemetry/api which provides no-op implementations when no SDK is registered.
 * withTrace() / @trace work safely without any OTEL setup; they just don't export spans.
 */

import {
  type Tracer,
  type Span,
  SpanStatusCode,
  trace,
  context as otelContext,
} from '@opentelemetry/api';
import { _emittedField, _incrementHealth } from './health';
import { getActiveOtelContext } from './propagation';
import { randomHex } from './hash';
import { shouldAllow } from './consent';
import { shouldSample } from './sampling';
import { tryAcquire, release } from './backpressure';

// Stryker disable next-line StringLiteral: tracer name is not observable without a real SDK
const TRACER_NAME = '@provide-io/telemetry';

// ── Manual trace context (injected without an active OTEL span) ───────────────
//
// Module-level globals are the fallback storage.  They are **not** safe when
// two overlapping async flows each call setTraceContext / withTrace without
// awaiting each other — one flow's state would leak into the other's Promise
// continuation.  When running on Node we reach for AsyncLocalStorage so each
// async chain gets its own isolated pair of IDs; the globals are reserved for
// non-scoped callers (direct setTraceContext usage) and for browser/Deno
// environments where AsyncLocalStorage is not available.
//
// Reads check the ALS store first, then fall back to the globals.  Writes via
// setTraceContext() prefer the ALS store when one is active (so a per-request
// scope wins over the globals), otherwise fall back to the globals (so the
// classic "set it once at boot" pattern still works).

type _TraceIds = { traceId?: string; spanId?: string };

let _als: {
  run<T>(store: _TraceIds, fn: () => T): T;
  getStore(): _TraceIds | undefined;
} | null = null;

try {
  // eslint-disable-next-line @typescript-eslint/no-require-imports
  const { AsyncLocalStorage } = require('node:async_hooks') as typeof import('node:async_hooks');
  _als = new AsyncLocalStorage<_TraceIds>();
} catch {
  // Non-Node runtime (browser, Deno): fall back to module globals.
  /* c8 ignore next */
  _als = null;
}

let _manualTraceId: string | undefined;
let _manualSpanId: string | undefined;

/**
 * Manually inject trace/span IDs (e.g. from an incoming request header).
 * Returned by getTraceContext(); cleared by _resetTraceContext().
 * Pass undefined (or empty string) for either argument to clear that ID.
 */
export function setTraceContext(traceId: string | undefined, spanId: string | undefined): void {
  // Treat empty strings as clearing the value — prevents empty-string IDs in logs.
  const normalizedTraceId = traceId || undefined;
  const normalizedSpanId = spanId || undefined;
  const store = _als?.getStore();
  if (store !== undefined) {
    store.traceId = normalizedTraceId;
    store.spanId = normalizedSpanId;
    return;
  }
  _manualTraceId = normalizedTraceId;
  _manualSpanId = normalizedSpanId;
}

/**
 * Return the current trace context: manual injection first, then active OTEL span.
 */
export function getTraceContext(): { trace_id?: string; span_id?: string } {
  const store = _als?.getStore();
  const traceId = store?.traceId ?? _manualTraceId;
  const spanId = store?.spanId ?? _manualSpanId;
  // Stryker disable next-line ConditionalExpression,LogicalOperator: setTraceContext always sets both; partial state not reachable via public API
  if (traceId !== undefined || spanId !== undefined) {
    return {
      // Stryker disable next-line ConditionalExpression: _manualTraceId is always defined here (both set together)
      ...(_manualTraceId !== undefined && { traceId: _manualTraceId }),
      // Stryker disable next-line ConditionalExpression: _manualSpanId is always defined here (both set together)
      ...(_manualSpanId !== undefined && { spanId: _manualSpanId }),
    };
  }
  const ids = getActiveTraceIds();
  return {
    ...(ids.trace_id !== undefined && { trace_id: ids.trace_id }),
    ...(ids.span_id !== undefined && { span_id: ids.span_id }),
  };
}

/** Reset manually injected trace context (used in tests). */
export function _resetTraceContext(): void {
  _manualTraceId = undefined;
  _manualSpanId = undefined;
  const store = _als?.getStore();
  if (store !== undefined) {
    store.traceId = undefined;
    store.spanId = undefined;
  }
}

/** Return the tracer for the telemetry library. Noop when no SDK is registered. */
export function getTracer(): Tracer {
  return trace.getTracer(TRACER_NAME);
}

/** Module-level lazy singleton tracer — resolves provider at call time. */
export const tracer: Tracer = trace.getTracer(TRACER_NAME);

/**
 * Return trace_id and span_id from the currently active OTEL span.
 * Returns an empty object when no span is active or OTEL is not configured.
 */
export function getActiveTraceIds(): { trace_id?: string; span_id?: string } {
  const span = trace.getActiveSpan();
  if (!span) return {};
  const ctx = span.spanContext();
  // OTEL no-op spans have all-zero IDs — treat as no active span.
  if (ctx.traceId === '00000000000000000000000000000000') return {};
  return { trace_id: ctx.traceId, span_id: ctx.spanId };
}

const NOOP_TRACE_ID = '00000000000000000000000000000000';

/**
 * Return true if the given span is a no-op (all-zero trace ID).
 * Used to decide whether to inject synthetic random IDs.
 */
function _isNoopSpan(span: Span): boolean {
  // Stryker disable next-line ConditionalExpression: without a registered OTel SDK all spans are noop — mutating to true is equivalent
  return span.spanContext().traceId === NOOP_TRACE_ID;
}

/**
 * Execute fn with random synthetic trace/span IDs set as manual context,
 * then restore the previous manual context when done.
 * Handles both sync and async results.
 */
function _withSyntheticIds<T>(fn: () => T): T {
  const prevTraceId = _manualTraceId;
  const prevSpanId = _manualSpanId;
  // Stryker disable next-line StringLiteral: random IDs are non-deterministic — exact value not observable in mutations
  setTraceContext(randomHex(16), randomHex(8));
  const result = fn();
  if (result instanceof Promise) {
    return result.then(
      (value) => {
        _manualTraceId = prevTraceId;
        _manualSpanId = prevSpanId;
        return value;
      },
      (err: unknown) => {
        _manualTraceId = prevTraceId;
        _manualSpanId = prevSpanId;
        throw err;
      },
    ) as T;
  }
  _manualTraceId = prevTraceId;
  _manualSpanId = prevSpanId;
  return result;
}

/** Shared span handler for withTrace — records exceptions and sets ERROR status. */
function _spanHandler<T>(fn: () => T, span: Span): T {
  _incrementHealth(_emittedField('traces'));
  try {
    const result = fn();
    if (result instanceof Promise) {
      return result.then(
        (value) => {
          span.end();
          return value;
        },
        (err: unknown) => {
          span.recordException(err instanceof Error ? err : new Error(String(err)));
          span.setStatus({ code: SpanStatusCode.ERROR, message: String(err) });
          span.end();
          throw err;
        },
      ) as T;
    }
    span.end();
    return result;
  } catch (err) {
    span.recordException(err instanceof Error ? err : new Error(String(err)));
    span.setStatus({ code: SpanStatusCode.ERROR, message: String(err) });
    span.end();
    throw err;
  }
}

/**
 * Execute fn inside a new span named `name`.
 * Works for both sync and async functions.
 * Mirrors Python @trace decorator behaviour: records exceptions, sets ERROR status.
 */
export function withTrace<T>(name: string, fn: () => T): T {
  if (!getConfig().tracingEnabled) return fn();
  // Stryker disable next-line StringLiteral: 'traces' vs '' is equivalent — shouldAllow treats any non-'logs'/non-'context' signal identically across all consent levels
  if (!shouldAllow('traces')) return fn();
  if (!shouldSample('traces', name)) return fn();
  const ticket = tryAcquire('traces');
  if (!ticket) return fn();

  const tracer = trace.getTracer(TRACER_NAME);

  // If an OTel context was extracted from propagation headers, use it as parent.
  // Narrow the try/catch to getActiveOtelContext() only so that errors inside
  // otelContext.with() propagate naturally instead of being swallowed and causing
  // a second span attempt in the fallback path without holding a backpressure slot.
  let activeCtx: ReturnType<typeof getActiveOtelContext> | undefined;
  try {
    const activeCtx = getActiveOtelContext();
    if (activeCtx) {
      try {
        return otelContext.with(activeCtx as ReturnType<typeof otelContext.active>, () =>
          tracer.startActiveSpan(name, (span: Span) => {
            // Stryker disable next-line ConditionalExpression: noop detection is not observable without SDK — branch outcome equivalent under mutation
            /* v8 ignore start: noop-span false branch + real-span return are unreachable without a registered OTel provider */
            if (_isNoopSpan(span)) return _withSyntheticIds(() => _spanHandler(fn, span));
            return _spanHandler(fn, span);
            /* v8 ignore stop */
          }),
        );
      } finally {
        release(ticket);
      }
    }
  } catch {
    // getActiveOtelContext() threw — graceful degradation, fall through to default behaviour.
  }

  try {
    return tracer.startActiveSpan(name, (span: Span) => {
      // Stryker disable next-line ConditionalExpression: noop detection is not observable without SDK — branch outcome equivalent under mutation
      if (_isNoopSpan(span)) return _withSyntheticIds(() => _spanHandler(fn, span));
      return _spanHandler(fn, span);
    });
  } finally {
    release(ticket);
  }
}

/**
 * Method/function decorator that wraps the target in withTrace().
 * Span name defaults to the decorated method name.
 *
 * Usage (requires experimentalDecorators: true):
 *   class Foo {
 *     @trace('my.operation')
 *     doWork() { ... }
 *   }
 */
// Stryker disable BlockStatement: outer and inner decorator fns returning undefined leave descriptors unchanged — equivalent
export function traceDecorator(name?: string) {
  return function (
    _target: object,
    propertyKey: string | symbol,
    descriptor: PropertyDescriptor,
  ): PropertyDescriptor {
    // Stryker enable BlockStatement
    // Stryker disable next-line LogicalOperator: span name not observable with no-op tracer
    const spanName = name ?? String(propertyKey);
    const original = descriptor.value as (...args: unknown[]) => unknown;
    descriptor.value = function (this: unknown, ...args: unknown[]) {
      return withTrace(spanName, () => original.apply(this, args));
    };
    return descriptor;
  };
}

File: go/tracing.go
// SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
// SPDX-License-Identifier: Apache-2.0

package telemetry

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"sync/atomic"
)

// Span represents an active trace span.
type Span interface {
	End()
	SetAttribute(key string, value any)
	RecordError(err error)
	SpanID() string
	TraceID() string
}

// Tracer creates and manages spans.
type Tracer interface {
	Start(ctx context.Context, name string) (context.Context, Span)
}

// _noopSpan is a no-op Span implementation used when no real tracer is configured.
type _noopSpan struct {
	traceID string
	spanID  string
}

func (s *_noopSpan) End()                               { _ = s }
func (s *_noopSpan) SetAttribute(key string, value any) { _ = key; _ = value }
func (s *_noopSpan) RecordError(err error)              { _ = err }
func (s *_noopSpan) SpanID() string                     { return s.spanID }
func (s *_noopSpan) TraceID() string                    { return s.traceID }

// _noopTracer is a no-op Tracer that generates random IDs and stores them in context.
type _noopTracer struct{}

func (t *_noopTracer) Start(ctx context.Context, name string) (context.Context, Span) {
	_ = name
	traceID := _randomHex(16)
	spanID := _randomHex(8)
	ctx = SetTraceContext(ctx, traceID, spanID)
	return ctx, &_noopSpan{traceID: traceID, spanID: spanID}
}

// _randomHex returns a random hex string of n bytes (2n hex chars).
func _randomHex(n int) string {
	b := make([]byte, n)
	_, _ = rand.Read(b)
	return hex.EncodeToString(b)
}

// _defaultTracer is the package-level tracer binding, defaulting to no-op.
//
// An atomic rather than an exported variable: it is read from every traced call
// and written during SetupTelemetry and ShutdownTelemetry, and a two-word
// interface value read while it is being written can tear into a stale itab
// paired with a new data pointer. An atomic load is ~1ns against the ~20ns a
// RWMutex costs on that path, and the same reasoning already applies to the
// signal gates in runtime.go.
//
// Reachable through [GetTracer] and replaceable through [SetDefaultTracer]; the
// spec's "tracer instance" symbol is satisfied by the exported [Tracer] type.
var _defaultTracer atomic.Pointer[Tracer] //nolint:gochecknoglobals

func _loadDefaultTracer() Tracer {
	if bound := _defaultTracer.Load(); bound != nil {
		return *bound
	}
	return _noopTracerSingleton
}

// _noopTracerSingleton is the binding before any setup, and after teardown.
var _noopTracerSingleton Tracer = &_noopTracer{} //nolint:gochecknoglobals

// SetDefaultTracer replaces the package-level tracer.
//
// Exported so a host can install its own implementation; SetupTelemetry and
// ShutdownTelemetry use it too. Safe to call concurrently with in-flight spans.
func SetDefaultTracer(t Tracer) {
	_defaultTracer.Store(&t)
}

// _traceIDKey and _spanIDKey are context keys for trace/span propagation.
var (
	_traceIDKey = contextKey{"trace.id"} //nolint:gochecknoglobals
	_spanIDKey  = contextKey{"span.id"}  //nolint:gochecknoglobals
)

// GetTracer returns a named Tracer. Currently returns the package-level tracer.
func GetTracer(name string) Tracer {
	_ = name
	return _loadDefaultTracer()
}

// Trace wraps fn in a span using the package-level tracer.
// fn receives the context enriched with trace/span IDs.
// If fn returns an error, the error is recorded on the span before it ends.
// Consent, sampling, and backpressure are applied before starting the span;
// fn is still invoked (without a span) when any gate rejects.
//
// When a live OTel tracer provider is installed, probabilistic sampling is
// delegated to the SDK ParentBased(TraceIDRatioBased) sampler so instrumented
// and facade spans share one sampling authority (no double-sampling).
// Without a live provider, ShouldSample(traces) applies as before.
func Trace(ctx context.Context, name string, fn func(context.Context) error) error {
	if !TracingEnabled() {
		return fn(ctx)
	}
	if !ShouldAllow(signalTraces, "") {
		return fn(ctx)
	}
	if !_hasLiveTraceProvider() {
		if sampled := _shouldSampleFailOpen(signalTraces, name); !sampled {
			return fn(ctx)
		}
	}
	ticket := TryAcquire(signalTraces)
	if ticket == nil {
		return fn(ctx)
	}
	defer Release(ticket)
	_incSpansStarted()

	spanCtx, span := _effectiveTracer().Start(ctx, name)
	defer span.End()
	err := fn(spanCtx)
	if err != nil {
		// Record exception to span; OTel spans forward this to the backend.
		span.RecordError(err)
	}
	return err
}

// _hasLiveTraceProvider reports whether a live tracer provider is in play — ours
// or a host application's — in which case the SDK's sampler is authoritative and
// the facade must not stack its own rate on top.
//
// Answered by the backend, which reports both what it installed and what it
// found on the OTel globals, and re-asked per span so a provider registered
// after setup counts (see _effectiveTracer and go/otel/adopt.go). The gate and
// the emit path therefore read the same state: a span is never sampled by the
// facade and then handed to an SDK sampler, nor vice versa.
func _hasLiveTraceProvider() bool {
	backend := _activeBackend()
	if backend == nil {
		return false
	}
	return backend.Providers().Traces
}

// GetTraceContext returns the trace and span IDs bound to ctx.
// When an active OTel span is present in ctx its IDs take precedence.
// Falls back to context key values set by SetTraceContext.
// Returns empty strings if not set.
func GetTraceContext(ctx context.Context) (traceID, spanID string) {
	if backend := _activeBackend(); backend != nil {
		if traceID, spanID, ok := backend.TraceContext(ctx); ok {
			return traceID, spanID
		}
	}
	if v, ok := ctx.Value(_traceIDKey).(string); ok {
		traceID = v
	}
	if v, ok := ctx.Value(_spanIDKey).(string); ok {
		spanID = v
	}
	return traceID, spanID
}

// SetTraceContext returns a new context with the given trace/span IDs bound.
func SetTraceContext(ctx context.Context, traceID, spanID string) context.Context {
	ctx = context.WithValue(ctx, _traceIDKey, traceID)
	ctx = context.WithValue(ctx, _spanIDKey, spanID)
	return ctx
}

// _getTraceSpanFromContext extracts trace/span IDs from context.
func _getTraceSpanFromContext(ctx context.Context) (traceID, spanID string) {
	return GetTraceContext(ctx)
}
